p2p/exchange_metrics.go

```diff
@@ -0,0 +1,165 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+var meter = otel.Meter("header/p2p")
+
+const (
+	failedKey           = "failed"
+	headerReceivedKey   = "num_headers_received"
+	headTypeKey         = "request_type"
+	headTypeTrusted     = "trusted_request"
+	headTypeUntrusted   = "untrusted_request"
+	headStatusKey       = "status"
+	headStatusOk        = "ok"
+	headStatusTimeout   = "timeout"
+	headStatusCanceled  = "canceled"
+	headStatusNoHeaders = "no_headers"
+)
+
+type exchangeMetrics struct {
+	headRequestTimeInst metric.Float64Histogram
+	responseSizeInst    metric.Int64Histogram
+	responseTimeInst    metric.Float64Histogram
+	blockedPeersNum     metric.Int64Counter
+
+	trackerPeersNum     atomic.Int64
+	trackedPeersNumInst metric.Int64ObservableGauge
+	trackedPeersNumReg  metric.Registration
+
+	disconnectedPeersNum     atomic.Int64
+	disconnectedPeersNumInst metric.Int64ObservableGauge
+	disconnectedPeersNumReg  metric.Registration
+}
+
+func newExchangeMetrics() (m *exchangeMetrics, err error) {
+	m = new(exchangeMetrics)
+	m.headRequestTimeInst, err = meter.Float64Histogram(
+		"hdr_p2p_exch_clnt_head_time_hist",
+		metric.WithDescription("exchange client head request time in seconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseSizeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_resp_size_hist",
+		metric.WithDescription("exchange client header response size in bytes"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseTimeInst, err = meter.Float64Histogram(
+		"hdr_p2p_exch_clnt_resp_time_hist",
+		metric.WithDescription("exchange client response time in seconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_trck_peer_num_gauge",
+		metric.WithDescription("exchange client tracked peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumReg, err = meter.RegisterCallback(m.observeTrackedPeers, m.trackedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_disc_peer_num_gauge",
+		metric.WithDescription("exchange client tracked disconnected peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumReg, err = meter.RegisterCallback(m.observeDisconnectedPeers, m.disconnectedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *exchangeMetrics) head(ctx context.Context, duration time.Duration, headersReceived int, tp, status string) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.headRequestTimeInst.Record(ctx,
+			duration.Seconds(),
+			metric.WithAttributes(
+				attribute.Int(headerReceivedKey, headersReceived),
```

Comment:
Is headersReceived falls into some predefined values?  If not, it is gonna be impossib;e to monitor amount of headers as attribute. It would be better as separate counter instead

Reply:
Its limited by 4 peers so yes